bin/helpers/openstack.sh

```diff
@@ -0,0 +1,357 @@
+  if [[ "${is_uploaded_disc_format_as_expected}" == "false" ]]; then
+    echo >&2 "Host image disc format: ${disc_format}"
+    echo >&2 "Server image disc format: ${server_image_disc_format}"
+    echo >&2 "${upload_name} disc format on the server is not as expected"
+    echo >&2 "Delete the server image and retry the kn init process"
+    exit 1
+  fi
+}
+
+function download_image() {
+  local download_filename=$1
+  local download_url=$2
+  local download_filepath="/tmp/${download_filename}"
+
+  echo "Downloading image to local /tmp/"
+  echo -e "\n"
+  curl "${download_url}/${download_filename}" \
+    -o "${download_filepath}" \
+    --connect-timeout 30 \
+    --max-time 1800
+
+  echo "Verifying the md5 checksum of the downloaded image"
+  echo -e "\n"
+  verify_downloaded_image "${download_filename}" "${download_url}"
+}
+
+function maybe_download_image(){
+  local image_name=$1
+  local disc_format=$2
+  local download_url=$3
+  local download_filename="${image_name}.qcow2"
+  local download_filepath="/tmp/${image_name}.qcow2"
+  local is_image_downloaded_already
+  is_image_downloaded_already=$(is_image_downloaded_already "${download_filepath}")
+
+  if [[ "${is_image_downloaded_already}" == "false" ]]; then
+    download_image "${download_filename}" "${download_url}"
+  else
+    echo "${download_filename} is already downloaded"
+    echo "Verifying the downloaded image checksum"
+    echo -e "\n"
+
+    verify_downloaded_image "${download_filename}" "${download_url}"
+  fi
+}
+
+function maybe_convert_image(){
+  local image_name=$1
+  local disc_format=$2
+  local download_url=$3
+  local download_filename="${image_name}.qcow2"
+  local download_filepath="/tmp/${image_name}.qcow2"
+  local upload_name="${image_name}"
+  local upload_filepath="/tmp/${image_name}.${disc_format}"
+  local qemu_installed
+  local is_image_downloaded_already
+  is_image_downloaded_already=$(is_image_downloaded_already "${download_filepath}")
+  qemu_installed=$(which qemu-img)
+
+
+  if [[ ! -e ${upload_filepath} ]]; then
+    if [[ "${qemu_installed}" == "" ]]; then
+      echo >&2 "qemu-utils package is missing - image cannot be converted"
+      echo >&2 "Please ensure that qemu-img is available."
+      exit 1
+    else
+      if [[ "${is_image_downloaded_already}" == "false" ]]; then
+        echo >&2 "No file found at ${download_filepath}"
+        echo >&2 "Please ensure image has been downloaded"
+        exit 1
+      fi
+
+      echo "${download_filepath} has not yet been converted to the ${disc_format} format"
+      echo "Starting the image conversion from qcow2 to ${disc_format}..."
+      echo -e "\n"
+
+      qemu-img convert -q -f qcow2 -O "${disc_format}" "${download_filepath}" "${upload_filepath}"
+    fi
+  else
+    echo "${upload_filepath} already exists and so does not require conversion, moving onto upload step"
+  fi
+}
+
+function maybe_upload_image() {
+  local image_name=$1
+  local disc_format=$2
```

Comment:
To avoid confusion with image format on image downloaded change "disk_format" to "glance_disk_format"